design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+
+Consider the following Smithy model:
+
+```smithy
+@length(min: 1, max: 69)
+string NiceString
+```
+
+The central idea to the implementation of constraint traits is: [parse, don't
+validate]. Instead of code-generating a Rust `String` to represent `NiceString`
+values and perform the _validation_ at request deserialization, we can
+[leverage Rust's type system to guarantee domain invariants]. We can generate a
+wrapper [tuple struct] that _parses_ the string's value and is "tight" in the
+set of values it can accept:
+
+```rust
+pub struct NiceString(String);
+
+impl NiceString {
+    pub fn parse(inner: String) -> Result<Self, NiceStringValidationError> {
+        Self::try_from(inner)
+    }
+}
+
+impl TryFrom<String> for NiceString {
+    type Error = NiceStringValidationError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        let num_code_points = value.chars().count();
+        if 1 <= num_code_points && num_code_points <= 69 {
+            Ok(Self(value))
+        } else {
+            Err(NiceStringValidationError::LengthViolation(num_code_points))
+        }
+    }
+}
+```
+
+(Note that we're using the _linear_ time check `chars().count()` instead of
+`len()` on the input value, since the Smithy specification says the `length`
+trait counts [the number of _Unicode code points_ when applied to string
+shapes].)
+
+The goal is to enforce, at the type-system level, that these constrained
+structs always hold valid data. It should be impossible for the service
+implementer, without resorting to `unsafe` Rust, to construct a `NiceString`
+that violates the model. The actual check is performed in the implementation of
+[`TryFrom`]`<InnerType>` for the generated struct, which makes it convenient to use
+the [`?` operator for error propagation]. Each constrained struct will have a
+related `std::error::Error` enum type to signal the _first_ parsing failure,
+with one enum variant per applied constraint trait:
+
+```rust
+pub enum NiceStringValidationError {
+    /// Validation error holding the number of Unicode code points found, when a value between `1` and
+    /// `69` (inclusive) was expected.
+    LengthViolation(usize),
+}
+
+impl std::error::Error for NiceStringValidationError {}
+```
+
+[`std::error::Error`] requires [`Display`] and [`Debug`]. We will
+`#[derive(Debug)]`, unless the shape also has the [`sensitive` trait], in which
+case we will just print the name of the struct:
+
+```rust
+impl std::fmt::Debug for NiceStringValidationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut formatter = f.debug_struct("NiceStringValidationError");
+        formatter.finish()
+    }
+}
+```
+
+`Display` is used to produce human-friendlier representations. Its
+implementation will be called when formatting a [400 HTTP response message] in
+certain protocols, for example. The format of the error messages will _not_ be
+configurable by service implementers and will be
```

Comment:
Preceding frameworks have allowed validation to be disabled or customized - from message all the way up to a completely different error. If a service implementer wanted to take control of the display of validation failures without opting out of the Rust SSDK validation framework entirely, how would they do so?

Reply:
This is the sort of thing that is much easier to do if you don't bake validation into the type system - in TS, handlers can be configured to accept a callback that is passed an array of `ValidationFailure` objects, and the user can turn those into an error of their choice that is then rendered by the framework. They still get the benefit of validation without giving up the presentation of modeled exceptions coming from their service.

Which reminds me of another point - we have largely done away with synthetic exceptions in Smithy. Validation is no difference - are you requiring that a service's operations be associated with a particular error structure?

Reply:
> in TS, handlers can be configured to accept a callback that is passed an array of ValidationFailure objects, and the user can turn those into an error of their choice that is then rendered by the framework.

Ah, I hadn't thought about that idea; I had imagined they'd want the errors to be configured ahead-of-time, not at runtime. That's neat. I don't see why we would not be able to build it into the Rust framework. We could have users return `struct`s (or functions that return `struct`s) that implement `std::error::Error` (or some trait of our own) for each of the `ValidationFailure`s in that callback, and we would use those representations when rendering the response.

> Which reminds me of another point - we have largely done away with synthetic exceptions in Smithy.

What are synthetic exceptions?

> are you requiring that a service's operations be associated with a particular error structure?

If the operation in the Smithy model is fallible (has an `errors` field), we require that the operation handlers return a [`Result<T, E>`](https://doc.rust-lang.org/std/result/enum.Result.html), where the error has to be the enum that we generate holding all the variants of their Smithy `errors` field.

Reply:
Synthetic exceptions are an idea from a legacy framework - exceptions that every service throws that don't have to be declared in the model. In that framework, `ValidationException` is synthetic, and can always be thrown. We'd like to restrict that idea to only things that are _unavoidably_ thrown, either by virtue of "this is how programming works" (`SerializationException`, `InternalFailure`) or the underlying protocol (incorrect values for `Accept`, etc).

In TS, you have two options:
- If every operation in the service has `smithy.framework#ValidationException` attached, and no customization is done in smithy-build.json, then the default validation behavior is engaged and the handlers cannot customize error rendering.
- If `smithy-build.json` has `disableDefaultValidation` set to `true`, code generation adds a **required** parameter to handlers for a callback that can turn `ValidationFailure[]` into an error (or undefined, if it wants to suppress the error)

If you don't attach `smithy.framework#ValidationException` to every operation in the service and you don't disable default validation in `smithy-build`, then code generation fails.

All this to say that you have to guarantee your validation failures result in modeled exceptions being rendered, and not unmodeled 400s.

Reply:
+1 for having the option to provide the function doing `ValidationFailure[] -> CustomErrorAlwaysMappedTo400`

Reply:
> you have to guarantee your validation failures result in modeled exceptions being rendered, and not unmodeled 400s.

I agree. Otherwise the server implementation breaks the Smithy contract. (I am actually curious about the current behavior if the input does not parse at all.)

As you mentioned, one approach to let the user choose which [error the operation will return](https://awslabs.github.io/smithy/1.0/spec/core/model.html#operation) is to generate for each operation input a type for validation error and ask the user to provide a function mapping this type to the operation error type.

Taking the [weather forecast example](https://awslabs.github.io/smithy/quickstart.html#complete-example) it would look like.

```rust
    enum GetCityInputError {
        CityId(CityIdValidationError)
    }

    enum CityIdValidationError {
        Pattern(String)
    }
```

A user would have to implement a mapping `fn(GetCityInputError) -> GetCityError`.

Reply:
@adamthom-amzn I didn't know about `ValidationException`, I don't see it documented in the docsite. Is the intention for this shape to be special, to be the default way for service frameworks to report validation errors?

> adds a required parameter to handlers for a callback that can turn ValidationFailure[] into an error

I'm assuming this error is modeled and associated to the operation. If the operation in the model does not define at least one error, code generation fails, right?

---

@oddg Would that mapping be static per operation (i.e. two identical invalid requests resulting in the same `GetCityError`)? Would it be a simple mapping (i.e. no access to state)?

If I were to implement this, I'm thinking that if default validation is disabled (e.g. in `smithy-build.json`) we could generate the operation registry to take in both the operation handler and the error mapping function per operation registration. Any other ideas as to where the user could provide the error mapping function?

Reply:
> Is the intention for this shape to be special, to be the default way for service frameworks to report validation errors?

Special? No, we're trying not to have anything be special. Think of it as a built-in convenience that service frameworks are free to use as a default :)

> I'm assuming this error is modeled and associated to the operation. If the operation in the model does not define at least one error, code generation fails, right?

This would be a fine behavior to have. Another is to just use your normal error serialization logic, which should turn any exception that is not associated with the operation into an `InternalFailure`.

> I am actually curious about the current behavior if the input does not parse at all.

@oddg by spec, here meaning protocol tests, unparseable input should result in a 400 with the error code of `SerializationException`. It's a synthetic, which as I outlined above, means non-modeled (and does not need to have a corresponding specialized type generated client-side).

Reply:
> Would that mapping be static per operation (i.e. two identical invalid requests resulting in the same GetCityError)? Would it be a simple mapping (i.e. no access to state)?

Yes. In my opinion, it makes sense that two identical invalid requests result in the same errors. I agree that it should not depend on any state. (Any error that is is state dependent should be generated in the core implementation of the handler for that operation.)

> We could generate the operation registry to take in both the operation handler and the error mapping function per operation registration.

That makes a lot of sense to me.

> by spec, here meaning protocol tests, unparseable input should result in a 400 with the error code of SerializationException. It's a synthetic, which as I outlined above, means non-modeled (and does not need to have a corresponding specialized type generated client-side).

Thanks.